controller/utils.go

```diff
@@ -122,104 +122,119 @@ func replacePodsAndWait(
 	goroutinePoolSize int,
 ) (timedout, canceled bool, err error) {
 	logger.Debug("starting to replace pods with new ones")
-	stop := make(chan struct{}, 1)
-	defer close(stop) // Dont leak goroutines
-	finishedReplace := make(chan struct{})
 
-	timedoutChan := make(chan bool)
-	canceledChan := make(chan bool)
+	ctx, cancel := context.WithCancel(context.Background())
+
+	finishedReplace := make(chan struct{})
+	canceledChan := make(chan struct{})
 	errChan := make(chan error)
 
 	pods := make(chan *models.Pod, len(podsChunk))
 	for _, pod := range podsChunk {
 		pods <- pod
 	}
 
+	inRollingUpdate := operationManager != nil
+
+	var wg sync.WaitGroup
 	logger.Infof("starting %d in-memory workers to replace %d pods", goroutinePoolSize, len(podsChunk))
 	for i := 0; i < goroutinePoolSize; i++ {
-		go replacePodWorker(
-			logger,
-			roomManager,
-			mr,
-			clientset,
-			db,
-			redisClient,
-			willTimeoutAt,
-			configYAML,
-			scheduler,
-			operationManager,
-			clock,
-			pods,
-			stop, finishedReplace,
-			timedoutChan, canceledChan, errChan,
-		)
+		wg.Add(1)
+		go func () {
+			defer wg.Done()
+			replacePodWorker(
+				ctx,
+				logger,
+				roomManager,
+				mr,
+				clientset,
+				db,
+				redisClient,
+				configYAML,
+				scheduler,
+				pods,
+				inRollingUpdate,
+				finishedReplace,
+				errChan,
+			)
+		}()
+	}
+	
+	if inRollingUpdate {
+		go func() {
+			for {
+				canceled, err := operationManager.WasCanceled()
```

Comment:
If this a busy loop? Adding a sleep on the goroutine is then maybe a good idea.

Reply:
When does this goroutine stop?

Reply:
Oh yes, I was going to put a ticker there but I forgot. I also have to use the context here too.